### PR TITLE
Adds modified versions of QueryPool, Query types from sigp discv5

### PR DESCRIPTION
**Pull request by mrferris**
### What does this add?

This PR adds relatively generic utility types used for iterative FindNodes/FindContent implementations, mostly taken from https://github.com/sigp/discv5/blob/master/src/query_pool.rs with some customizations/re-organization done. 

Customizations include:

- Stripping out of `PredicateQuery` types. These allowed one to specify an arbitrary filter to run over ENRs discovered. Removing this type is the reason for the compiler currently requiring `PhantomData` fields in `query_pool.rs`, because the `PredicateQuery` definition used the `TResult` field while the `FindNodeQuery` definition does not.
- Modify `QueryInfo` to use `Trin` types when generating requests instead of `sigp/discv5' types.

The main types defined are the `QueryPool` type, the `Query` type (both defined in `find/query_pool.rs`), and the `FindNodeQuery`  type (defined in `find/iterators/findnodes.rs`). 

The `QueryPool` type holds a set of `Query` entries, each of which has an instance of the `QueryPeerIter` enum, which points to a `FindNodeQuery` (or a `FindContentQuery` once implemented), whose `on_success`/`on_failure` methods define the rules by which the query is advanced each time a response is received from a peer. 

### How will it be used?

The subsequent PR that uses these types:

-  adds a `QueryPool` member to the `OverlayService` struct.
-  adds a function that polls the `QueryPool` for `QueryPoolState` objects, based off of which the query will either finish or a`FINDNODES` message will be sent to the next peer (whose info is contained in the `QueryPoolState` object) via a `TALK_REQ`.
- adds a function that calls `on_success`/`on_failure` when a `NODES` message is received from a node that is part of a query.
- calls the `QueryPool`'s `add_findnode_query` method in the OverlayService to begin a FindNodes query.

Another subsequent PR will need to define `FindContentQuery` and its respective `on_success`/`on_failure` methods, at which point this approach will work for `FindContent` as well. I envision the `OverlayService` having two `QueryPool` members, one for each type of query. Each `QueryPool` will be instantiated with different types corresponding to the types of data that will be returned (content or ENRs).

### CODE REVIEW

File: trin-core/src/portalnet/find/query_pool.rs
Comment (jacobkaufmann) [inline trin-core/src/portalnet/find/query_pool.rs:272]: Have you considered how we would extend this to results for `FindContent` queries, where the result may not contain a list of closest peers?

Comment (jacobkaufmann) [inline trin-core/src/portalnet/find/query_pool.rs:272]: Have you considered how we would extend this to results for `FindContent` queries, where the result may not contain a list of closest peers?

File: trin-core/src/portalnet/find/query_pool.rs
Comment (jacobkaufmann) [inline trin-core/src/portalnet/find/query_pool.rs:199]: Related to comment below. This implementation seems to imply that the result type will be some key for a node ID. This is also evident in the `on_success` method with the type of the `new_peers` argument.

File: trin-core/src/portalnet/find/query_pool.rs
Comment (jacobkaufmann) [inline trin-core/src/portalnet/find/query_pool.rs:41]: Why not use `QueryId` type here?

File: trin-core/src/portalnet/find/query_info.rs
Comment (jacobkaufmann) [inline trin-core/src/portalnet/find/query_info.rs:67]: The `log2_distance` method relies on [XOR distance](https://github.com/sigp/discv5/blob/master/src/kbucket/key.rs#L91) defined on `Key` in `sigp/discv5`. This is not a problem for the history network, but this may require some rework for the state network.

File: trin-core/src/portalnet/find/iterators/findnodes.rs
Comment (jacobkaufmann) [inline trin-core/src/portalnet/find/iterators/findnodes.rs:388]: Why do we have `iteration` field? It looks like this is not a field for `QueryPeer` on the master branch of `sigp/discv5`. Also, we seem to never use this field.

**Review by jacobkaufmann — COMMENTED** [review]
A change like this is hard to evaluate on its own so most of my comments focus on extensibility to other queries like `FindContent`.

Also, it looks like there was a minor discrepancy between the implementation of the `FindNodeQuery` here and the one in `sigp/discv5`, so I would like to confirm that we based this PR off the latest from the master branch there.

I don't have any changes to request, but I would like to wait until the comments are addressed before approval.

File: trin-core/src/portalnet/find/iterators/findnodes.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/find/iterators/findnodes.rs:284]: *nitpick* `count` > `cnt`

File: trin-core/src/portalnet/find/iterators/findnodes.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/find/iterators/findnodes.rs:590]: And this

File: trin-core/src/portalnet/find/query_info.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/find/query_info.rs:44]: Why not `TNodeId` to maintain consistency with the `findnodes.rs` code?

**Review by njgheorghita — APPROVED** [review]
Lookin' good! The comments are super helpful. Mostly nitpicks from me, a bit hard to judge the api in isolation, but i'm looking forward to the following prs & findcontent support

File: trin-core/src/portalnet/find/iterators/findnodes.rs
Comment (jacobkaufmann) [reply trin-core/src/portalnet/find/iterators/findnodes.rs:140]: IIUC, this is true for the XOR distance metric, but not for the state distance function.

File: trin-core/src/portalnet/find/query_pool.rs
Comment (mrferris) [reply trin-core/src/portalnet/find/query_pool.rs:272]: I'm thinking that we could add a `retrieved_content: Option<Vec<u8>>` field to this struct.

File: trin-core/src/portalnet/find/query_pool.rs
Comment (mrferris) [reply trin-core/src/portalnet/find/query_pool.rs:199]: Similar to above, I picture us adding an optional content field to `on_success`.

As for `TResult`, that's a good point, but in my mind I don't see us needing to use `TResult` to return the content if we pass in a channel to send the content back over. In this case, we might be able to get rid of `TResult` altogether. It was mostly used for the predicate queries, which have been removed.

Comment (mrferris) [reply trin-core/src/portalnet/find/query_pool.rs:199]: Similar to above, I picture us adding an optional content field to on_success.

As for TResult, that's a good point, but in my mind I don't see us needing to use TResult to return the content if we pass in a channel to send the content back over. In this case, we might be able to get rid of TResult altogether. It was mostly used for the predicate queries, which have been removed.

File: trin-core/src/portalnet/find/query_info.rs
Comment (mrferris) [reply trin-core/src/portalnet/find/query_info.rs:67]: good point, i'll file an issue after i merge this

File: trin-core/src/portalnet/find/iterators/findnodes.rs
Comment (mrferris) [reply trin-core/src/portalnet/find/iterators/findnodes.rs:151]: Good point @njgheorghita. Added a note. After thinking about it, I think it's extremely unlikely that it does happen (with XOR) but even if it does, it doesn't look like the results would be very severe if presented with results from a node that it thinks it already got results from. It would just return, so basically one of the nodes would be chosen and the other ignored.

Good point @jacobkaufmann. I'll add that to the issue I file about making queries distance function configurable.

**Pull request merged**
